File: 2096.py
# 2096. Step-By-Step Directions From a Binary Tree Node to Another
# Definition for a binary tree node.
from typing import Optional, List

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def getDirections(self, root: Optional[TreeNode], startValue: int, destValue: int) -> str:
        def getPath(cur, targetValue, path, ans):
            if cur is None:
                return
            if cur.val == targetValue:
                ans.append(''.join(path))
            path.append('L')
            getPath(cur.left, targetValue, path, ans)
            path[-1] = 'R'
            getPath(cur.right, targetValue, path, ans)
            path.pop(-1)

        tmpPath = []
        startPath = []
        destPath = []
        getPath(root, startValue, tmpPath, startPath)
        getPath(root, destValue, tmpPath, destPath)
        startPath = startPath[0]
        destPath = destPath[0]

        i = 0
        while i < min(len(startPath), len(destPath)) and startPath[i] == destPath[i]:
            i += 1
        return 'U' * (len(startPath) - i) + destPath[i:]

def build_tree(values: List[Optional[int]]) -> Optional[TreeNode]:
    if not values:
        return None
    root = TreeNode(values[0])
    queue = [root]
    i = 1
    while i < len(values):
        current = queue.pop(0)
        if values[i] is not None:
            current.left = TreeNode(values[i])
            queue.append(current.left)
        i += 1
        if i < len(values) and values[i] is not None:
            current.right = TreeNode(values[i])
            queue.append(current.right)
        i += 1
    return root

# Example binary tree
#     5
#    / \
#   1   2
#  /   / \
# 3   6   4

values = [5, 1, 2, 3, None, 6, 4]
root = build_tree(values)

# Create an instance of the Solution class
solution = Solution()

# Test case 1: Path from node 3 to node 6
startValue = 3
destValue = 6
result = solution.getDirections(root, startValue, destValue)
print(f"Path from {startValue} to {destValue}: {result}")  # Expected: "UURL"

# Test case 2: Path from node 6 to node 4
startValue = 6
destValue = 4
result = solution.getDirections(root, startValue, destValue)
print(f"Path from {startValue} to {destValue}: {result}")  # Expected: "UR"

# Add more test cases as needed

File: 2096.js
//  Definition for a binary tree node.
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

// @param {TreeNode} root
// @param {number} startValue
// @param {number} destValue
// @return {string}

var getDirections = function (root, startValue, destValue) {
  function getPath(cur, targetValue, path, result) {
    if (!cur) return;
    if (cur.val === targetValue) {
      result.push(path.join(""));
      return;
    }
    path.push("L");
    getPath(cur.left, targetValue, path, result);
    path.pop();
    path.push("R");
    getPath(cur.right, targetValue, path, result);
    path.pop();
  }

  let startPath = [];
  let destPath = [];
  let tmpPath = [];

  // Find paths from root to startValue and destValue
  getPath(root, startValue, tmpPath, startPath);
  getPath(root, destValue, tmpPath, destPath);

  startPath = startPath[0];
  destPath = destPath[0];

  // Find the common prefix
  let i = 0;
  while (
    i < Math.min(startPath.length, destPath.length) &&
    startPath[i] === destPath[i]
  ) {
    i++;
  }

  // Return the result with 'U's for going up and remaining part of destPath
  return "U".repeat(startPath.length - i) + destPath.slice(i);
};

// Example usage:

// Build the binary tree
function buildTree(values) {
  if (!values.length) return null;
  const root = new TreeNode(values[0]);
  const queue = [root];
  let i = 1;
  while (i < values.length) {
    const node = queue.shift();
    if (values[i] !== null) {
      node.left = new TreeNode(values[i]);
      queue.push(node.left);
    }
    i++;
    if (i < values.length && values[i] !== null) {
      node.right = new TreeNode(values[i]);
      queue.push(node.right);
    }
    i++;
  }
  return root;
}

// Example binary tree
//     5
//    / \
//   1   2
//  /   / \
// 3   6   4

const values = [5, 1, 2, 3, null, 6, 4];
const root = buildTree(values);

console.log(getDirections(root, 3, 6)); // Expected output: "UURL"
console.log(getDirections(root, 6, 4)); // Expected output: "UR"
